Fix hostname and address uniqueness validation for hosts sharing an IP

kke opened the pull request:
Fixes #1439

Hostname uniqueness check would ignore hosts with the same IP address but a different SSH port.

A host with a non-uniqu address but a different SSH port would have been considered a duplicate, eventhough it's likely a port forward.

jakolehm commented on lib/pharos/phases/validate_host.rb:
```diff
@@ -44,10 +44,10 @@ def check_role
 
       def validate_unique_hostnames
-        duplicates = @config.hosts.reject { |h| h.address == @host.address }.select { |h| h.hostname == @host.hostname }
+        duplicates = @config.hosts.reject { |h| h.address == @host.address }.select { |h| h.hostname == @host.hostname && h.ssh_port == @host.ssh_port }
```
Hmm... why are we calling this unique hostnames and then we validate unique addresses? 🤔

kke replied:
`GatherFacts` does `host.hostname = transport.exec("hostname -f")`.

If it's a port forward to some other host, it should get a different hostname anyway. 

Maybe the unique hostname thing actually needs to just check for duplicate hostnames, no matter what the address / port is, so something like `(@config.hosts - [@host]).find { |h| h.hostname == @host.hostname }` ?

kke replied:
Hmm, in fact, I think instead of reporting hostnames as duplicate, it should have been allowing duplicate hostnames for hosts with the same address?

`@config.hosts.reject { |h| h.address == @host.address }` would have been empty if there were two hosts with the same address but different ssh ports.

jakolehm replied:
We are mixing two different things here:
1. uniqueness of address:port which is used for ssh
2. uniqueness of hostname which is used to identify nodes in kubernetes

kke replied:
config_schema will complain if addresses are equal and does not check ssh_port.

Needs to be fixed also.

kke commented:
Added host:ssh_port uniqueness validation and removed the unique address validation.

Had to resort to trickery in `ConfigSchema#load` because couldn't think of a way to do it with a dry predicate.

jakolehm commented on lib/pharos/config_schema.rb:
```diff
@@ -34,6 +34,16 @@ def self.load(data)
       result = schema.call(DEFAULT_DATA.merge(data))
       raise Pharos::ConfigError, result.messages unless result.success?
 
+      messages = { hosts: {} }
```
IMO this can be implemented as predicate

```ruby
    def self.build
      # rubocop:disable Lint/NestedMethodDefinition
      Dry::Validation.Params do
        configure do

          # @param hosts [Array<Hash>]
          def unique_addresses?(hosts)
            return false if hosts.group_by { |host| "#{host[:address]}:#{host[:ssh_port]}" }.size < hosts.size

            true
          end
```

```ruby
required(:hosts).filled(:unique_addresses?, min_size?: 1) do
```

kke commented:
Now uses a predicate.

- It does not tell which host is non-unique
- The message is displayed also when the `hosts:` is missing.
- When it fails, it always makes `size cannot be less than 1` appear also, I suppose it clears the "non valid" `hosts:` and then decides it's too small.

kke commented on spec/pharos/phases/validate_host_spec.rb:
```diff
@@ -150,12 +150,27 @@
 
-        expect{ subject.validate_unique_hostnames }.to raise_error(Pharos::InvalidHostError, "Duplicate hostname foo for hosts 192.0.2.2,192.0.2.3")
+        it 'raises if duplicates' do
+          expect{ subject.validate_unique_hostnames }.to raise_error(Pharos::InvalidHostError, "Duplicate hostname foo for hosts 192.0.2.2:22,192.0.2.3:22")
```
Hmm, how can this pass? This sort of message is not there anymore

kke replied:
Ah, it's in the phase.
